server/otlp/http_server.go

```diff
@@ -0,0 +1,222 @@
+package otlp
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	pb "go.opentelemetry.io/proto/otlp/collector/trace/v1"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+)
+
+const protoBufContentType = "application/x-protobuf"
+const jsonContentType = "application/json"
+const gzipEncoding = "gzip"
+
+type HttpServer struct {
+	addr     string
+	exporter IExporter
+
+	hServer *http.Server
+}
+
+func NewHttpServer(addr string, exporter IExporter) *HttpServer {
+	return &HttpServer{
+		addr:     addr,
+		exporter: exporter,
+	}
+}
+
+func (s *HttpServer) Start() error {
+	s.hServer = &http.Server{
+		Addr:    s.addr,
+		Handler: s,
+	}
+
+	listener, err := net.Listen("tcp", s.addr)
+	if err != nil {
+		return fmt.Errorf("cannot listen on address %s: %w", s.addr, err)
+	}
+
+	return s.hServer.Serve(listener)
+}
+
+func (s *HttpServer) Stop() {}
+
+func (s HttpServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	contentType, contentEncoding, err := s.validateRequest(r)
+	shouldDecompress := contentEncoding == gzipEncoding
+	shouldEncode := strings.Contains(r.Header.Get("Accept-Encoding"), gzipEncoding)
+
+	response := newHttpResponse(w, contentType, shouldEncode)
+	if err != nil {
+		response.sendError(http.StatusBadRequest, status.Errorf(codes.InvalidArgument, "Error when validating request %s", err.Error()))
+		return
+	}
+
+	request, err := s.parseBody(r.Body, contentType, shouldDecompress)
+	if err != nil {
+		response.sendError(http.StatusUnprocessableEntity, status.Errorf(codes.InvalidArgument, "Could not parse request body %s", err.Error()))
+		return
+	}
+
+	result, err := s.exporter.Ingest(r.Context(), request)
+	if err != nil {
+		response.sendError(http.StatusInternalServerError, status.Errorf(codes.InvalidArgument, "Error when ingesting spans %s", err.Error()))
+		return
+	}
+	response.send(http.StatusOK, result)
+}
+
+func (s HttpServer) parseProtoBuf(body []byte) (*pb.ExportTraceServiceRequest, error) {
+	request := pb.ExportTraceServiceRequest{}
+
+	err := proto.Unmarshal(body, &request)
+	if err != nil {
+		return nil, err
+	}
+
+	return &request, nil
+}
+
+func (s HttpServer) parseJson(body []byte) (*pb.ExportTraceServiceRequest, error) {
```

Comment:
Ideally, we should be able to support both protobuf and JSON body protocols.

The problem is that the defined struct for the incoming request https://github.com/open-telemetry/opentelemetry-proto-go/blob/main/otlp/collector/trace/v1/trace_service.pb.go#L48 requires the resource spans in a JSON to be inside a field called `resource_spans` but the node exporters like  https://www.npmjs.com/package/@opentelemetry/exporter-trace-otlp-http send the data in a different format. This is the format we get:

```
{
  "resourceSpans": [
    {
      "resource": {
        "attributes": [
          {
            "key": "service.name",
            "value": { "stringValue": "unknown_service:node" }
          },
          {
            "key": "telemetry.sdk.language",
            "value": { "stringValue": "nodejs" }
          },
          {
            "key": "telemetry.sdk.name",
            "value": { "stringValue": "opentelemetry" }
          },
```

We are using `json.Unmarshal` and as those fields don't match we are not getting the correct data.
A similar scenario is happening with tempo and can be found here 

https://github.com/kubeshop/tracetest/blob/6a203b23bce8924dc3fc6a1929647fb0ec396aee/server/traces/otel_http_converter.go